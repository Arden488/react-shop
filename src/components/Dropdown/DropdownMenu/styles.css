.dropdown-menu {
  background: #fff;
  box-shadow: 0 1px 10px 0 rgba(0, 0, 0, .1);
  text-align: initial;
}

.dropdown-menu a {
  display: block;
  padding: 2px 4px;
  text-decoration: none;
  /* color: #000; */
}

.dropdown-menu a:hover {
  background-color: rgba(0, 0, 0, .1);
}
