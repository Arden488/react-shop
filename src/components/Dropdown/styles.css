.dropdown {
  position: relative;
}

.dropdown__trigger button {
  /* display: block; */
  cursor: pointer;
}

.dropdown__menu,
.dropdown__content {
  position: absolute;
  top: 100%;
  min-width: 100%;
  opacity: 0;
  visibility: hidden;
}

.dropdown__menu_active,
.dropdown__content_active {
  opacity: 1;
  visibility: visible;
}
