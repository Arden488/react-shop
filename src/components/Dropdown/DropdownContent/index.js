import React, { Component } from 'react';

import './styles.css';

class DropdownContent extends Component {
  render() {
    return (
      <div className="dropdown-content">
        {this.props.children}
      </div>
    );
  }
}

export default DropdownContent;
