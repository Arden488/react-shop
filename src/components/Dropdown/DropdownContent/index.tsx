import * as React from 'react';

import './styles.css';

class DropdownContent extends React.Component {
  public render() {
    return (
      <div className="dropdown-content">
        {this.props.children}
      </div>
    );
  }
}

export default DropdownContent;
