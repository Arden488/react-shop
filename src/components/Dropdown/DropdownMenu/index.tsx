import React, { Component } from 'react';

import './styles.css';

class DropdownMenu extends Component {
  render() {
    return (
      <div className="dropdown-menu">
        {this.props.children}
      </div>
    );
  }
}

export default DropdownMenu;
