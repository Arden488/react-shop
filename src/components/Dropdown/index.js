import React, { Component } from 'react';
import PropTypes from 'prop-types';

import DropdownMenu from './DropdownMenu';

import './styles.css';

class Dropdown extends Component {
  constructor(props) {
    super(props);

    this.content = [];
    this.menu = '';

    this.state = {
      openDropdownMenu: false,
    };

    this.prefilterChildren(this.props.children);

    this.handleTriggerClick = this.handleTriggerClick.bind(this);
    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  prefilterChildren(children) {
    React.Children.forEach(children, (child) => {
      if (child.type === DropdownMenu) {
        this.menu = child;
      } else {
        this.content.push(child);
      }
    });
  }

  setWrapperRef(node) {
    this.wrapperRef = node;
  }

  handleClickOutside(event) {
    console.log(this.wrapperRef.contains(event.target));
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({
        openDropdownMenu: false,
      });
    }
  }

  handleTriggerClick() {
    this.setState({
      openDropdownMenu: true,
    });
  }

  render() {
    let menuClasses = 'dropdown__menu';
    if (this.state.openDropdownMenu) menuClasses += ' dropdown__menu_active';

    return (
      <div className="dropdown" ref={this.setWrapperRef}>
        <div className="dropdown__trigger" onClick={this.handleTriggerClick}>
          {this.content}
        </div>
        <div className={menuClasses}>
          {this.menu}
        </div>
      </div>
    );
  }
}

Dropdown.propTypes = {
  children: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default Dropdown;
