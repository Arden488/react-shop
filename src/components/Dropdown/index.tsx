/* global document */

import * as React from 'react';

import DropdownContent from './DropdownContent/';
import DropdownMenu from './DropdownMenu/';

import './styles.css';

// tslint:disable-next-line
interface Dropdown {
  wrapperRef: any,
  triggerContent: any[],
  content: React.ReactChild,
  menu: React.ReactChild,
}

interface IDropdownState {
  openDropdownMenu: boolean,
}

class Dropdown extends React.Component<any, IDropdownState> {
  constructor(props: any) {
    super(props);

    this.triggerContent = [];
    this.content = '';
    this.menu = '';

    this.state = {
      openDropdownMenu: false,
    };

    this.prefilterChildren(this.props.children);

    this.handleTriggerClick = this.handleTriggerClick.bind(this);
    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this.renderMenu = this.renderMenu.bind(this);
    this.renderContent = this.renderContent.bind(this);
  }

  public render() {
    return (
      <div className="dropdown" ref={this.setWrapperRef}>
        <div className="dropdown__trigger" onClick={this.handleTriggerClick}>
          {this.triggerContent}
        </div>
        {this.menu && this.renderMenu()}
        {this.content && this.renderContent()}
      </div>
    );
  }

  public componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  public componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  public handleTriggerClick() {
    this.setState({
      openDropdownMenu: !this.state.openDropdownMenu,
    });
  }

  private setWrapperRef(node: any) {
    this.wrapperRef = node;
  }

  private prefilterChildren(children: React.ReactNode) {
    React.Children.forEach(children, (child: React.ReactElement<any> ) => {
      if (child.type === DropdownMenu) {
        this.menu = child;
      } else if (child.type === DropdownContent) {
        this.content = child;
      } else {
        this.triggerContent.push(child);
      }
    });
  }

  private handleClickOutside(event: { target: any }) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({
        openDropdownMenu: false,
      });
    }
  }

  private renderMenu() {
    let menuClasses = 'dropdown__menu';
    if (this.state.openDropdownMenu) { menuClasses += ' dropdown__menu_active'; }

    return (
      <div className={menuClasses}>
        {this.menu}
      </div>
    );
  }

  private renderContent() {
    let contentClasses = 'dropdown__content';
    if (this.state.openDropdownMenu) { contentClasses += ' dropdown__content_active'; }

    return (
      <div className={contentClasses}>
        {this.content}
      </div>
    );
  }
}

export default Dropdown;
