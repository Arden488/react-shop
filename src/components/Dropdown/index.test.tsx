import { shallow } from 'enzyme';
import * as React from 'react';

import DropdownContent from './DropdownContent';
import DropdownMenu from './DropdownMenu';
import Dropdown from './index';

let component: any = null;

beforeEach(() => {
  component = shallow(<Dropdown><button /></Dropdown>);
});

it('matches snapshot', () => {
  expect(component).toMatchSnapshot();
});

it('matches snapshot with menu', () => {
  component = shallow(<Dropdown><button /><DropdownMenu /></Dropdown>);
  expect(component).toMatchSnapshot();
});

it('matches snapshot with content', () => {
  component = shallow(<Dropdown><button /><DropdownContent /></Dropdown>);
  expect(component).toMatchSnapshot();
});

it('should have trigger with onclick event handler', () => {
  expect(component.find('.dropdown__trigger').prop('onClick')).toBeDefined();
});

it('should have openDropdownMenu state', () => {
  expect(component.state('openDropdownMenu')).toBeDefined();
});

it('should have trigger that handle click events and change state', () => {
  const spy = jest.spyOn(Dropdown.prototype, 'handleTriggerClick');
  component = shallow(<Dropdown><button /></Dropdown>);

  expect(component.state('openDropdownMenu')).toBeFalsy();

  component.find('.dropdown__trigger').simulate('click');
  expect(spy).toHaveBeenCalled();
  expect(component.state('openDropdownMenu')).toBeTruthy();
  spy.mockRestore();
});
