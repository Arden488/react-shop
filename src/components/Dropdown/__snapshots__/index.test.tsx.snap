// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="dropdown"
>
  <div
    className="dropdown__trigger"
    onClick={[Function]}
  >
    <button />
  </div>
</div>
`;

exports[`matches snapshot with content 1`] = `
<div
  className="dropdown"
>
  <div
    className="dropdown__trigger"
    onClick={[Function]}
  >
    <button />
  </div>
  <div
    className="dropdown__content"
  >
    <DropdownContent />
  </div>
</div>
`;

exports[`matches snapshot with menu 1`] = `
<div
  className="dropdown"
>
  <div
    className="dropdown__trigger"
    onClick={[Function]}
  >
    <button />
  </div>
  <div
    className="dropdown__menu"
  >
    <DropdownMenu />
  </div>
</div>
`;
